### ISSUE TITLE
[BUG] Multiple Copies of an array breaks the computational graph

### ISSUE THREAD

**Issue by dyumanaditya** [NVIDIA/warp#717]
### Bug Description

Hello,

## What I'm trying to do
My objective is to

1.  start with a warp array called `base_state`,
2.  make `n` copies of this array,
3. perform some operations on each copy,
4. average the resulting arrays and 
5. perform backpropagation

In the code attached below, I 

1. start with an array `base_state = [0.0,  0.5, 1.0]`
2. make `num_samples` copies of this array
3. in a loop over `num_samples` multiply one of the copied arrays by 2, then by 4. Each intermediate operation is stored so that the gradient can be computed and there are no array overwrites.
4. Average the final arrays and sum to compute a scalar loss

From my understanding the expected behavior is that the final gradient should be `[8. 8. 8.]` regardless of the number of samples.

<details>
  <summary>Click to expand — Code</summary>

```python
# minimal_repro_warp.py
import warp as wp
import numpy as np
import time

# wp.config.verify_autograd_array_access=True

# 0) initialize Warp
wp.init()


@wp.kernel
def multiply_by_2(
        x: wp.array(dtype=wp.float32),
        out: wp.array(dtype=wp.float32)  # write-only
):
    tid = wp.tid()
    if tid < x.shape[0]:
        out[tid] = x[tid] * 2.0


@wp.kernel
def multiply_by_4(
        x: wp.array(dtype=wp.float32),
        out: wp.array(dtype=wp.float32)  # write-only
):
    tid = wp.tid()
    if tid < x.shape[0]:
        out[tid] = x[tid] * 4.0


@wp.kernel
def avg_scalars(
    buf: wp.array2d(dtype=wp.float32),  # shape: (num_samples, elements)
    n: wp.int32,                        # number of samples
    out: wp.array(dtype=wp.float32)     # write-only, length = idx.shape[0]
):
    i, sample = wp.tid()
    if i < buf.shape[1]:
        wp.atomic_add(out, i, buf[sample, i] / wp.float32(n))


@wp.kernel
def sum_to_loss(
    avg_state: wp.array(dtype=wp.float32),
    loss: wp.array(dtype=wp.float32)     # write-only, length = 1
):
    tid = wp.tid()
    wp.atomic_add(loss, 0, avg_state[tid])


elements = 3
num_samples = 10
base_state = wp.array(np.linspace(0.0, 1.0, elements, dtype=np.float32), requires_grad=True)
loss = wp.zeros(1, dtype=wp.float32, requires_grad=True)
print("base_state: ", base_state)
print()

intermediate_states = [[wp.zeros(elements, dtype=wp.float32, requires_grad=True) for _ in range(2)] for _ in range(num_samples)]
tape = wp.Tape()

with tape:
    nom_states = [wp.zeros_like(base_state) for _ in range(num_samples)]
    # nom_states = [base_state for _ in range(num_samples)]
    for s in range(num_samples):
        # nom_state = wp.clone(base_state, requires_grad=True)

        # NOTE: This doesn't work
        nom_state = wp.zeros_like(base_state, requires_grad=True)
        wp.copy(nom_state, base_state)

        # NOTE: This works, but I would be modifying the base_state which is not ideal
        # nom_state = base_state


        wp.launch(
            multiply_by_4,
            dim=elements,
            inputs=[nom_state],
            outputs=[intermediate_states[s][0]],
        )

        # perform some operation on the perturbed state
        wp.launch(
            multiply_by_2,
            dim=elements,
            inputs=[intermediate_states[s][0]],
            outputs=[intermediate_states[s][1]],
        )

    # average the results
    # 1) allocate a 2D buffer: (num_samples x elements)
    buf = wp.zeros(
        (num_samples, elements),
        dtype=wp.float32,
        requires_grad=True
    )

    # 3) output averaged vector
    avg_state = wp.zeros(elements, dtype=wp.float32, requires_grad=True)

    # 4) gather each sample's post-op into one row of buf
    for s in range(num_samples):
        wp.copy(buf[s], intermediate_states[s][1])

    # 5) launch averaging kernel
    wp.launch(
        avg_scalars,
        dim=(elements, num_samples),
        inputs=[buf, num_samples],
        outputs=[avg_state]
    )

    # sum to scalar loss
    wp.launch(
        sum_to_loss,
        dim=elements,
        inputs=[avg_state],
        outputs=[loss]
    )

print("avg_state:", avg_state)
print("loss:", loss)
tape.backward(loss)
print("∂Loss/∂base_state:", base_state.grad)
```

</details>

## What occurs
The gradient seems to be scaled by a factor of `1/num_samples` (ex. if `num_samples=10` then the gradient becomes `[0.8, 0.8, 0.8]`)

If, however, instead of copying, I set `nom_state = base_state`, the gradient is computed correctly, but this would mean modifying the base state which is not ideal.

Any help on this would be appreciated, I've attached a simple schematic describing the flow as well

![Image](https://github.com/user-attachments/assets/3dcbc383-b196-4220-86dc-b92242eb7ea6)

 

### System Information

Warp 1.7.0
CUDA Toolkit 12.8, Driver 12.8
OS: Ubuntu 24.04
